server/src/main/java/org/opensearch/extensions/action/ExtensionTransportActionsHandler.java

```diff
@@ -120,19 +121,26 @@ public TransportResponse handleTransportActionRequestFromExtension(TransportActi
                 @Override
                 public void onResponse(ExtensionActionResponse actionResponse) {
                     response.setResponseBytes(actionResponse.getResponseBytes());
-                    inProgressLatch.countDown();
+                    inProgressFuture.complete(actionResponse);
                 }
 
                 @Override
                 public void onFailure(Exception exp) {
                     logger.debug("Transport request failed", exp);
                     byte[] responseBytes = ("Request failed: " + exp.getMessage()).getBytes(StandardCharsets.UTF_8);
                     response.setResponseBytes(responseBytes);
-                    inProgressLatch.countDown();
+                    inProgressFuture.completeExceptionally(exp);
                 }
             }
         );
-        inProgressLatch.await(ExtensionsManager.EXTENSION_REQUEST_WAIT_TIMEOUT, TimeUnit.SECONDS);
+        try {
+            inProgressFuture.orTimeout(ExtensionsManager.EXTENSION_REQUEST_WAIT_TIMEOUT, TimeUnit.SECONDS).join();
+        } catch (CompletionException e) {
+            if (e.getCause() instanceof TimeoutException) {
+                logger.info("No response from extension to request.");
+            }
+            throw e;
```

Comment:
Please rethow `e.getCause()` - the `CompletionException` is just a wrapper, it has no value for the caller but the real cause has.

Reply:
I meant to change that to a log I think.  The problem with throwing e.getCause() is that it's a throwable, so we have to add throws declarations to every method.

Reply:
That is correct, so we have to check if this is runtime exception (free to rethrow) or needs wrapping

Reply:
In that scenario, we would have to cast it as an exception right?

Reply:
Correct, if this is not a `RuntimeException` or Error, we could cast it to Exception and rethrow